### PR TITLE
US001 - View Job Roles

### PR DESCRIPTION
**Pull request by BlankAmber**
# Pull Request Checklist

Please follow all these rules before merging your branch.
- [x] Must reference a ticket
- [x] Must implement tests where appropriate
- [x] Must have useful documentation
- [x] Must adhere to the naming conventions
- [x] Must not commit confidential information 

Once all requirements above are fulfilled, your pull request has no merge conflicts, and you have two approvals from the approved approvers, then it can be tested and signed off.

# PR dependencies
- #12 

# Relevant tickets
- Implements frontend for [US001 - View Job Roles](https://tasks.office.com/kainos.com/Home/Task/3UAGcsT_y0uD2Mv9Bt5HlZYAMd3i?Type=TaskLink&Channel=Link&CreatedTime=638386268083560000

A JobRoleController, JobRoleService and JobRole model has been added to process and display Job Role details from the back end api. The list-JobRole.html file takes data which is passed from the controller and displays it in a table.

Note: screenshots and video are slightly out of date as some minor CSS changes have been made
Screenshot 2023-12-18 at 15 42 25

Pagination has been applied throughout these files to default limit the amount of results displayed in the html file to 50. 2 buttons have been added to the bottom of the list to allow the user to change the data shown. These 2 buttons have hover effects, to increase in size so the user can easily tell that it is a pressable button. When the user can not go back or forwards a page, that button will turn grey, and the hover effect will be disabled so the user can easily tell that they can no long press it. 

Screenshot 2023-12-14 at 16 22 18

https://github.com/BlankAmber/Kainos-RRAS-UI/assets/145463825/ecad77ac-bed8-4bd7-9f88-3c8f16e894ca

https://github.com/BlankAmber/Kainos-RRAS-UI/assets/145463825/16d17289-b011-451c-a7f9-2a46d8d9829e

A drop down button has been added, to allow the user to select how many job roles they want to appear on each page. the can select 25, 50, 100 or all. 

Screenshot 2023-12-20 at 14 15 48

https://github.com/BlankAmber/Kainos-RRAS-UI/assets/145463825/bd1e8651-d63a-4038-a38f-4b6b24ce1ff6

If the back-end can not be connected too, then an error message will display telling the user that there are no job roles detected.

Screenshot 2023-12-21 at 09 34 49

Images have been added to the public folder to meet Kainos branding requirements (US026 - Add branding to the product) Branding applied includes, Kainos logo, Kainos colours, styling and fonts, Home page, Footer with contact information.

Screenshot 2023-12-14 at 16 22 25
Screenshot 2023-12-14 at 16 22 38
Screenshot 2023-12-21 at 09 15 20

# Changes
Fixes to tidy up code and fix inconsistent naming conventions

All html <script> code has been moved out of html and into a separate js file

minor css fixes

### CODE REVIEW

**Comment by lullabelle** [comment]
gorgeous pull request!

File: model/JobRole.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export class JobRole {
+    jobRoleId?: number;
+    jobRoleName?: string;
+    JobRole: string;
```
Comment (jkelso662) [inline model/JobRole.ts:4]: How come the capitalisation is different here?

File: controller/jobRoleController.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import { Application, Request, Response } from "express";
+import { JobRoleServiceResponse } from "../service/jobRoleServiceResponse";
+
+const jobRoleService = require("../service/jobRoleService");
+
+module.exports = function(app: Application) {
+    app.get("/JobRoles", async(req: Request, res: Response) => {
```
Comment (jkelso662) [inline controller/jobRoleController.ts:7]: If the `/login` endpoint is lowercase, why is it `/JobRole`? Would have at least expected it to be `/jobRole`

File: controller/jobRoleController.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            totalResults: 0
+        };
+        const numOfResults = 50;
+
+        try {
+            data = await jobRoleService.getJobRoles(req.session.token, numOfResults, numOfResults);
+        } catch (e) {
+            console.error(e);
+        }
+        res.render("list-JobRole", { JobRoles: data.firstResults, totalResults: data.totalResults });
```
Comment (jkelso662) [inline controller/jobRoleController.ts:19]: kebab-case plus PascalCase hurts my brain! haha

File: views/list-JobRole.html [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+            }
+
+            let newNumber = parseInt(lastResult, 10);
+
+            if(newNumber >= jobRolesTotalResults){
+                document.getElementById("loadMoreButton").disabled = true;
+                document.getElementById("loadMoreButton").classList.add("disabled");
+            }else{
+                document.getElementById("loadMoreButton").disabled = false;
+                document.getElementById("loadMoreButton").classList.remove("disabled");
```
Comment (jkelso662) [inline views/list-JobRole.html:119]: Is there any reason why this is not in a js or ts file but is instead in the HTML? Same question to the other bits and pieces scattered in HTML.

**Review by jkelso662 — DISMISSED** [review]
Really good work on getting the pagination working and well done on including a video of it working. Nice touch. Would have been good to see the error pages when you run out of items. 

I did however have a couple of things worth looking at if you wouldn't mind. Thanks!

**Comment by BlankAmber** [comment]
TODO: @adammarshall745 bug fix

**Review by ryanmcalary11 — APPROVED** [review]
Great work! 😁

File: model/jobRole.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+export class JobRole {
+    jobRoleId: number;
+    jobRoleName: string;
+    JobRole: string;
```
Comment (lullabelle) [inline model/jobRole.ts:4]: was this supposed to be changed to jobRole?

Comment (rionamcglinchey) [reply model/jobRole.ts:4]: > was this supposed to be changed to jobRole?

(This is Samuel): I think Josh only requested the file name to be changed from JobRole to jobRole for consistency. The other classes all use camel case with the first letter capitalised

Comment (jkelso662) [reply model/jobRole.ts:4]: Can this reference to jobRole not also be corrected to match jobRoleId and jobRoleName here?

Comment (rionamcglinchey) [reply model/jobRole.ts:4]: > Can this reference to jobRole not also be corrected to match jobRoleId and jobRoleName here?

(Samuel): OH I didn't even notice line 4 that's my bad

Comment (lullabelle) [reply model/jobRole.ts:4]: can you fix it pls? then we will approve

Comment (rionamcglinchey) [reply model/jobRole.ts:4]: > can you fix it pls? then we will approve

(Samuel) Someone took my laptop...

Comment (adammarshall745) [reply model/jobRole.ts:4]: Fixed @jkelso662 @lullabelle

**Review by lullabelle — APPROVED** [review]

**Pull request merged**